### PR TITLE
Follow-up #12035 and #12068: Show an informative progress indicator when dragging and dropping PDF files to JabRef 

### PR DESCRIPTION
**Pull request by BaronMac08**
Closes [#8738 ](https://github.com/JabRef/jabref/issues/8738)
A follow-up to #12035 and #12068
Builds upon: the branch `fix-melting-630` merged at [#12001](https://github.com/JabRef/jabref/pull/12001)

Modifications:
- Dragging and dropping several PDF files into JabRef will be one large import rather than one small import for every file

Additions:
- Added informative messages about the ongoing progress of the big import by building upon the BackgroundTask progress indicator
- Added shortenFileName(fileName, maxLength) in StringUtil to shorten and prominently display the files being imported
- Added one new unit test for validating StringUtil.shortenFileName()

UI Screenshots after modifications and additions:
Screenshot 2024-10-24 at 20 09 55
Screenshot 2024-10-24 at 20 11 11
Screenshot 2024-10-24 at 20 22 29
Screenshot 2024-10-24 at 20 22 37

### Mandatory checks

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -40,6 +40,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - We added automatic browser extension install on Windows for Chrome and Edge. [#6076](https://github.com/JabRef/jabref/issues/6076)
 - We added a search bar for filtering keyboard shortcuts. [#11686](https://github.com/JabRef/jabref/issues/11686)
 - By double clicking on a local citation in the Citation Relations Tab you can now jump the linked entry. [#11955](https://github.com/JabRef/jabref/pull/11955)
+- When dragging and dropping several PDF files into the main table, the menu icon for background tasks becomes a progress indicator visualising the import's progress. [#12072](https://github.com/JabRef/jabref/pull/12072)
```
Comment (mag-sun) [inline CHANGELOG.md:43]: Maybe explicitly say you added a progress indicator i.e. "We added a progress indicator that visualises an import's progress after dragging and dropping several PDF files...." etc

File: src/main/java/org/jabref/model/strings/StringUtil.java [hunk 2]
```diff
@@ -167,6 +170,66 @@ public static String getCorrectFileName(String orgName, String defaultExtension)
...
+
+        StringBuilder result = new StringBuilder();
+        result.append(name, 0, numCharsBeforeEllipsis)
+              .append(ELLIPSIS)
+              .append(name.substring(name.length() - numCharsAfterEllipsis))
+              .append(extension);
+
+        return result.toString();
+    }
+
```
Comment (mag-sun) [inline src/main/java/org/jabref/model/strings/StringUtil.java:232]: Your code is clear, and handles edge cases well.

File: src/test/java/org/jabref/model/strings/StringUtilTest.java [hunk 3]
```diff
@@ -367,6 +368,28 @@ void capitalizeFirst() {
...
+            "thisisatestfile.pdf            , thisisatestfile.pdf                 , 20",
+            "...                            , longfilename.extremelylongextension ,  5",
+            "...                            , longfilename.extremelylongextension , 10",
+            "...                            , longfilename.extremelylongextension , 20",
+            "lo...me.extremelylongextension , longfilename.extremelylongextension , 30",
+    })
+    void shortenFileName(String expected, String fileName, int maxLength) {
+        assertEquals(expected, StringUtil.shortenFileName(fileName, maxLength));
+    }
+
```
Comment (mag-sun) [inline src/test/java/org/jabref/model/strings/StringUtilTest.java:392]: Potentially add a test for `null` inputs for file name?

File: src/main/java/org/jabref/model/strings/StringUtil.java (hunk 2 above)
Comment (BaronMac08) [reply src/main/java/org/jabref/model/strings/StringUtil.java:232]: Thanks @mag-sun!

File: src/test/java/org/jabref/model/strings/StringUtilTest.java (hunk 3 above)
Comment (BaronMac08) [reply src/test/java/org/jabref/model/strings/StringUtilTest.java:392]: Definitely! Using the @CsvSource, I have already added some tests with null fileNames. In the CSV-format strings, the first field is the expected string, the second is the inputted fileName, and lastly is the maxLength
Screenshot 2024-10-25 at 21 11 44

However, thanks to your suggestion, I identified another potential null case, which is the maxLength, and will also add tests for this case.

Thanks @mag-sun !

File: CHANGELOG.md (hunk 1 above)
Comment (BaronMac08) [reply CHANGELOG.md:43]: I will refactor the comment to be more precise and follow the existing comments. Thank you!

**Comment by BaronMac08** [comment]
Peer review from @mag-sun completed, the PR is now ready for review.

**Comment by BaronMac08** [comment]
Hi @koppor, I have opened this PR as you requested in #12068. For the two failing unit tests:
1. StringUtil should have less than or equal to 774 lines of code 
2. doNotUseLogicInModel was triggered due to the usage of FileUtil within StringUtil

Please let me know if there are any changes you would like me to make, thank you in advance for the feedback.

**Comment by koppor** [comment]
@BaronMac08 https://github.com/JabRef/jabref/pull/12068 was closed. Maybe you mean https://github.com/JabRef/jabref/pull/12072 -- please let pull requests open after changing something. Otherwise, it is really hard to follow.

**Comment by koppor** [comment]
> 2\. doNotUseLogicInModel was triggered due to the usage of FileUtil within StringUtil

I think, not. `import org.jabref.logic.util.io.FileUtil;` is in logic, too. -- Please recheck.

**Comment by koppor** [comment]
> 1. StringUtil should have less than or equal to 774 lines of code

Move the code to `FileUtil`. I think, the cohesion is better there.

**Comment by BaronMac08** [comment]
I have moved the shortenFileName() method to FileUtil, as @koppor suggested. All checks have passed.

File: src/test/java/org/jabref/logic/util/io/FileUtilTest.java [hunk 4]
```diff
@@ -477,4 +478,28 @@ void legalPaths(String fileName) {
...
+            "''                             , thisisatestfile.pdf                 ,   ",
+            "''                             , thisisatestfile.pdf                 , -5",
+            "''                             , thisisatestfile.pdf                 ,  0",
+            "...                            , thisisatestfile.pdf                 ,  3",
+            "...                            , thisisatestfile.pdf                 ,  5",
+            "th...e.pdf                     , thisisatestfile.pdf                 , 10",
+            "thisisatestfile.pdf            , thisisatestfile.pdf                 , 20",
+            "...                            , longfilename.extremelylongextension ,  5",
+            "...                            , longfilename.extremelylongextension , 10",
+            "...                            , longfilename.extremelylongextension , 20",
```
Comment (koppor) [inline src/test/java/org/jabref/logic/util/io/FileUtilTest.java:499]: This is not good behavior.

Expected:

```suggestion
            "lo..                          , longfilename.extremelylongextension ,  5",
            "longfil...                   , longfilename.extremelylongextension , 10",
            "longfilename.extr....                            , longfilename.extremelylongextension , 20",
```

Reason: There are plenty !! of characters available at 10 and 20. They should be used. Not just 3

File: src/test/java/org/jabref/logic/util/io/FileUtilTest.java [hunk 5]
```diff
@@ -477,4 +478,28 @@ void legalPaths(String fileName) {
...
+            "''                             ,                                     , -3",
+            "''                             ,                                     ,  0",
+            "''                             ,                                     ,  3",
+            "''                             ,                                     ,  5",
+            "''                             ,                                     , 10",
+            "''                             , thisisatestfile.pdf                 ,   ",
+            "''                             , thisisatestfile.pdf                 , -5",
+            "''                             , thisisatestfile.pdf                 ,  0",
+            "...                            , thisisatestfile.pdf                 ,  3",
+            "...                            , thisisatestfile.pdf                 ,  5",
```
Comment (koppor) [inline src/test/java/org/jabref/logic/util/io/FileUtilTest.java:494]: Please! Use all 5 characters: `th...`.

**Review by koppor — CHANGES_REQUESTED** [review]
Small comments on the functionality of the ellipses. Maybe, you can fix it?

**Comment by koppor** [comment]
Additional comment:

The number of files is displayed multiple times:

image

Please change progress information before the pipe sign to `Importing files into 12072.bib |`

And if you are on it, can you also change it for the indexing string?

image

**Comment by koppor** [comment]
Drop of files does not work. ALL files are attached to ALL entries. Even on pressing Cancel.

image

Please try out your things. Having 500 PDFs available is not too hard. I used all from my Downloads folder to test. - If not, you can try out the PDFs available at https://github.com/koppor/tugboat-mirror/tree/main/tugboat_issues.

File: src/test/java/org/jabref/logic/util/io/FileUtilTest.java (hunk 5 above)
Comment (BaronMac08) [reply src/test/java/org/jabref/logic/util/io/FileUtilTest.java:494]: Definitely, I will modify the functionality to make use of the available characters.

**Comment by BaronMac08** [comment]
> Drop of files does not work. ALL files are attached to ALL entries. Even on pressing Cancel.
> 
> image
> 
> Please try out your things. Having 500 PDFs available is not too hard. I used all from my Downloads folder to test. - If not, you can try out the PDFs available at https://github.com/koppor/tugboat-mirror/tree/main/tugboat_issues.

Thank you for pointing out the issue. I believe the issue is caused by the grouping of all files' entries and then sending it to the indexManager as one big EntriesAddedEvent. Therefore, I perform the exact same test of dragging and dropping the PDF files provided at [https://github.com/koppor/tugboat-mirror/tree/main/tugboat_issues](https://github.com/koppor/tugboat-mirror/tree/main/tugboat_issues) into `JabRef/main`  and branch `fix-8738` with the following results:

1. On `JabRef/main`:
Screenshot 2024-10-27 at 10 13 20
Screenshot 2024-10-27 at 10 15 45

2. On `fix-8738`:
Screenshot 2024-10-27 at 10 03 25
Screenshot 2024-10-27 at 10 02 45

The difference between `main` and `fix-8738` for the entry "... To Don Knuth..." in the linked files section:
1. On `main`: tb139complete.pdf and tb138complete.pdf are at the beginning of the list
2. On `fix-8738`: tb138complete.pdf and tb139completed.pdf are at the end of the list

Additionally, on both `main` and `fix-8738`, the file "The Communications of the TEX Users Group.pdf" has the same linked file.

I am wondering what would be the correct behaviour for this dataset of PDF files so that the issue on branch `fix-8738` could be identified and resolved. Thank you in advance for the clarification @koppor.

**Review by koppor — APPROVED** [review]

**Comment by koppor** [comment]
> I am wondering what would be the correct behaviour for this dataset of PDF files so that the issue on branch `fix-8738` could be identified and resolved.

Thank you for investigating. I needed to raise a new issue at https://github.com/JabRef/jabref/issues/12098.

**Comment by koppor** [comment]
Thank you for the quick follow-up in the filename methods. This is good to go now!

**Comment by BaronMac08** [comment]
Thank you for your guidance and advice!
